### ISSUE TITLE
error: does not create index

### ISSUE THREAD

**Issue by mikedfunk** [m2mdas/phpcomplete-extended#24]
When I open a composer project it detects it and says generating autload classmap... then generating index... then I get this screen:

![screen shot 2014-02-15 at 4 00 41 am](https://f.cloud.github.com/assets/661038/2177797/1b6e9b66-963a-11e3-88a8-a553ed85fa6e.png)

When I reopen vim in that directory it tries to build the index again as if it's never been built. There is no file created called .phpcomplete_extended/phpcomplete_index. I've set the .phpcomplete_extended dir to 777 recursively. When I do `:PHPCompleteExtendedGenerateIndex --verbose` I get no errors, yet the index file is not created. My `.phpcomplete_extended` dir contains these files:

core_index
index_cache

Why is it not generating the index?

**Comment by m2mdas** [comment]
Sorry for late reply.

That is odd. The only case when the indexer script does not create `phpcomplete_index` file is due to failure of `json_encode`. Can you put `echo` around [this](https://github.com/m2mdas/phpcomplete-extended/blob/master/bin/IndexGenerator.php#L61) section to see that if there is any error in json encode?

**Comment by Charl13** [comment]
Yesterday I encountered the same problem trying to generate a index for  a PHP roject using Zend Framework 2. Steps to reproduce the problem (using a small ZF2 project) are:

`curl -s https://getcomposer.org/installer | php --`
`php composer.phar create-project -sdev --repository-url="https://packages.zendframework.com" zendframework/skeleton-application path/to/install`
`cd path/to/install`
`vim`

After that it will try to create an index asking `Composer project detected, Do you want to create index? [yes/no]:`

After I type `yes` it shows me:

`PHP Deprecated:  Query route deprecated as of ZF 2.1.4; use the "query" option of the HTTP router's assembling method instead in /var/www/path/to/install/vendor/zendframework/zendframework/library/Zend/Mvc/Router/Http/Query.php on line 22`
`Deprecated: Query route deprecated as of ZF 2.1.4; use the "query" option of the HTTP router's assembling method instead in /var/www/path/to/install/vendor/zendframework/zendframework/library/Zend/Mvc/Router/Http/Query.php on line 22`
`- Malformed UTF-8 characters, possibly incorrectly encoded`
`Loading Index`
`Error detected while processing function phpcomplete_extended#readDataForProject..phpcomplete_extended#loadProject..<SNR>54_loadIndex..<SNR>54_readIndex:`
`line    2:`
`Could not read index file .phpcomplete_extended/phpcomplete_index`
`Index Loaded.`
`Press ENTER or type command to continue`

**Comment by Charl13** [comment]
Hmm it does seems to be a JSON parsing error. After downloading IndexGenerator.php to the root dir of my project, I could manage to get some output by issuing this command `php -f IndexGenerator.php generate -verbose`.

It processes alot of file but it stops on various php files with the output `- Malformed UTF-8 characters, possibly incorrectly encoded`.

**Comment by mikedfunk** [comment]
Good idea @Bertabuffel . I did a similar thing but didn't get as clear output. It is failing on IndexGenerator for certain files [here](https://github.com/m2mdas/phpcomplete-extended/blob/master/bin/IndexGenerator.php#L506). This prevents it from continuing completely and halts execution with no error displayed in stderr.

So 2 things: Ideally this needs better error checking to see if `execHook()` fails (or on a lower level?) and the error needs to be fixed in the first place.

**Comment by mikedfunk** [comment]
Update:

I updated my php install. I'm now getting a maximum nesting level reached from xdebug. It's on [IndexGenerator on line 744](https://github.com/m2mdas/phpcomplete-extended/blob/master/bin/IndexGenerator.php#L744). Hopefully that helps.

**Comment by m2mdas** [comment]
@Bertabuffel, thank you for detailed steps for reproducing the problem, I am currently looking into it. 

@mikedfunk, Your problem seems to be a major bug in the project. Can you post some detailed information about your project?

**Comment by m2mdas** [comment]
By major bug I actually meant bug in my project :)

**Comment by m2mdas** [comment]
@Bertabuffel, I have fixed the bug. Please update the plugin and run `:PHPCompleteExtendedGenerateIndex` again to see if problem still persists. Also you will not get auto-completion of all classes as ZF2 implements a custom autoloader last time I checked. Hoping to create a ZF2 extension in this weekend.

@mikedfunk , can you update the plugin and run `:PHPCompleteExtendedGenerateIndex --verbose` command again? After that can you go to the last class it processed before the xdebug error occured and issue `:g/\(^use\|class\)/p` and paste the output here. I am assuming there is a bug in my class parser code.

**Comment by mikedfunk** [comment]
I am working on a large project for work with hundreds of classes. I have specified an `include-path` array in composer.json. When I commented out [lines 744-747](https://github.com/m2mdas/phpcomplete-extended/blob/master/bin/IndexGenerator.php#L744-L747) I was able to continue with trying to generate the index without recursion. `IndexGenerator::validateClass` calls itself infinitely from those lines. From there phpcomplete-extended found some syntax errors lying around and find an invalid path for a `require` statement. Once I fixed those it worked.

**Comment by mikedfunk** [comment]
@m2mdas I did that and this time it failed on Predis. Your command yielded use statements, class declarations, _and comments_. I assume you didn't intend for it to match comments. For instance, this was matched:

```
  * Base class for Redis commands.
```

... which was part of a docblock. It also matched some // comments. There was a ton of output so I didn't paste it all here.

**Comment by m2mdas** [comment]
Well I actually needed the lines till class declaration to check the structure. Nevermind, can you link which file of Predis is causing the error?

**Comment by mikedfunk** [comment]
I will post it later, I'm heading out the door at the moment. Btw I am still getting the recursion problem  with the latest version unless I comment out those lines.

**Comment by m2mdas** [comment]
Ok, no problem.

**Comment by mikedfunk** [comment]
[Here is the full class](http://hastebin.com/kawexomagu.php). Do you still need the output of `:g/\(^use\|class\)/p`?

**Comment by Charl13** [comment]
@M2mdas Great I will update the plugin and run the command. Will post the results asap!

**Comment by m2mdas** [comment]
@mikedfunk, Hmm it seems that the file have multiple class. But the plugin currently supports one class per file. Due to this limitation recursion is occurring it seems. To support multiple files per class semi-major redesign is needed. I will implement the feature I need some free time to test all cases. In the meantime if the problem goes away by commenting out the lines you mentioned previously you can create a fork and change the lines needed and update from there. I won't be doing major change in the plugin soon. Sorry for this inconvenience.

**Comment by MichaelMackus** [comment]
I'm also getting a PHP Fatal Error during index generation in a Sylius application. 

Steps to reproduce:

composer create-project -s dev sylius/sylius-standard acme

Then triggering an index generation fails with "PHP Fatal error:  Cannot redeclare class AuthorizeNetRequest"

If I comment out the multiple requires in AuthorizeNet.php it gets past that class, but then fails with "PHP Fatal error:  Declaration of FOS\RestBundle\Request\RequestBodyParamConverte
r::apply() must be compatible with Sensio\Bundle\FrameworkExtraBundle\Request\Pa
ramConverter\ParamConverterInterface::apply(Symfony\Component\HttpFoundation\Req
uest $request, Sensio\Bundle\FrameworkExtraBundle\Configuration\ConfigurationInt
erface $configuration)"

I think the best way for this to work would be to generate the index regardless of fatal errors. Not sure the best way to approach that, but its frustrating that this project runs fine in the web, but I can't get phpcomplete to index it properly.
